File: src/api/resource.ts
import { RequestFN } from "~/core/request-function";
import { decodeResource } from "~/decoders/resource";
import { TabLocation, type Resource, type SessionHandle } from "~/models";

export const resource = async (session: SessionHandle, resourceID: string): Promise<Resource> => {
  const request = new RequestFN(session, "donneesContenusCDT", {
    _Signature_: { onglet: TabLocation.Resources },

    donnees: {
      cahierDeTextes: { N: resourceID }
    }
  });

  const response = await request.send();

  const resource = response.data.donnees.ListeCahierDeTextes.V[0];
  return decodeResource(resource, session);
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/api/resource.kt
package ink.literate.pawnote.api

import ink.literate.pawnote.core.RequestFN
import ink.literate.pawnote.decoders.decodeResource
import ink.literate.pawnote.models.Resource
import ink.literate.pawnote.models.SessionHandle
import ink.literate.pawnote.models.TabLocation
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

suspend fun resource(session: SessionHandle, resourceID: String): Resource {
  val request =
      RequestFN(
          session.information,
          "donneesContenusCDT",
          Json.encodeToString(
              buildJsonObject {
                putJsonObject("_Signature_") { put("onglet", TabLocation.Resources.code) }

                putJsonObject("donnees") {
                  putJsonObject("cahierDeTextes") { put("N", resourceID) }
                }
              }))

  val response = request.send()

  val resource =
      Json.parseToJsonElement(response.data)
          .jsonObject["donnees"]!!
          .jsonObject["ListeCahierDeTextes"]!!
          .jsonObject["V"]!!
          .jsonArray[0]
  return decodeResource(resource.jsonObject, session)
}
